ipa-core/src/protocol/ipa_prf/boolean_ops/multiplication.rs

```diff
@@ -1,33 +1,164 @@
-            addition_sequential::integer_add,
-        },
+        basics::mul::SecureMul, boolean::NBitStep, context::Context,
+        ipa_prf::boolean_ops::addition_sequential::integer_add, BooleanProtocols, Gate, RecordId,
     },
     secret_sharing::{replicated::semi_honest::AdditiveShare, BitDecomposed, FieldSimd},
 };
 
-
+/// This function multiplies x by y in these steps:
+/// 1. Double the input precision for y and repeat the most significant bit in the extra bits (Sign extension)
+///    X is assumed to be a positive number, so we will automatically pad with ZERO.
+/// 2. Repeatedly multiply x with each digits of y, shift the result 1 digit up each time
+/// 3. Add up the partial products using `integer_add`
+/// x is assumed to be a positive number
+/// y is assumed to be in two's complement and can be either signed or unsigned
+#[allow(dead_code)]
 pub async fn integer_mul<C, S, const N: usize>(
     ctx: C,
     record_id: RecordId,
     x: &BitDecomposed<AdditiveShare<Boolean, N>>,
     y: &BitDecomposed<AdditiveShare<Boolean, N>>,
-) -> Result<
-    BitDecomposed<AdditiveShare<Boolean, N>>,
-    Error,
->
+) -> Result<BitDecomposed<AdditiveShare<Boolean, N>>, Error>
 where
     C: Context,
     S: NBitStep,
     Boolean: FieldSimd<N>,
     AdditiveShare<Boolean, N>: BooleanProtocols<C, N>,
+    Gate: StepNarrow<S>,
 {
-    //TODO: To be implemented
+    use super::step::MultiplicationStep as Step;
+
+    let new_len = x.len() + y.len();
+    let mut y = y.clone();
+    y.resize(new_len, y[y.len() - 1].clone());
+
+    let mut result = BitDecomposed::with_capacity(new_len);
+    for (i, yb) in y.into_iter().enumerate() {
+        let ctx_for_bit_of_y = ctx.narrow(&S::from(i));
+        let product_of_x_and_yb = ctx_for_bit_of_y
+            .parallel_join(x.iter().take(new_len - i).enumerate().map(|(j, xb)| {
+                let ctx_for_x_times_y_combo = ctx_for_bit_of_y.narrow(&S::from(j));
+                let yb = yb.clone();
+                async move { yb.multiply(xb, ctx_for_x_times_y_combo, record_id).await }
+            }))
+            .await?;
+
+        let t = BitDecomposed::try_from(product_of_x_and_yb).unwrap();
+        if i == 0 {
+            result = t;
+        } else {
+            // add up bits i.. with the product
+            let mut add_y = BitDecomposed::with_capacity(new_len);
+            result.iter().skip(i).for_each(|b| add_y.push(b.clone()));
+            let (add_result, carry) = integer_add::<_, S, N>(
+                ctx_for_bit_of_y.narrow::<Step>(&Step::Add),
+                record_id,
+                &t,
+                &add_y,
+            )
+            .await?;
+
+            result.truncate(i); // replace the bits on i.. with the one we add up
+            add_result.iter().for_each(|b| result.push(b.clone()));
+            if result.len() < new_len {
+                // if carry bit is more than max length, we let it overflow
+                result.push(carry);
```

Comment:
I am pretty sure that this line will never be called. Looking at the Google sheet, I think it's the case that result has the maximum length from the very first line.

Reply:
the intermediate `result` won't have the green part in the Google sheet. I added this because I don't extend x's size anymore (it'll always be padded with 0s). But in that case we have to consider the carry bit from `integer_add` when we add the intermediate products, otherwise there are cases where  it'll drop the carry and err by 1 bit.

Reply:
Very cool. Makes sense!